README.md

```diff
@@ -1,37 +1,78 @@
+Generators are used to creates source files from some other resource file. For example take `.proto` files and run `protoc` on them to generate their respectice `.h` and `.cc` files. Or running a custom script to create a lookup table from a text file.
+
+### Type Generators
+A type generator instructs `cppdep` how to tranform a filetype into other files that will be used as part of the compile process. Each type generator config requires that three config parameters be set:
+
+* `inputext` : `string` -- the extension of the filetype on which to use this generator. This should include the `.`.
+* `outputexts`: `array of string` -- the extension(s) of the resulting file(s) from the generations process.
+* `command`: `array of string` -- a command and arguments to be run in order to to transform the input file into output files.
+
+#### Further Details 
+This command is not run in a shell so environment variables will not be available except for a few special variables that will be substituted by `cppdep` itself. The variables provided are as follows:
+* `$CPPDEP_INPUT_DIR`: the path to the directory in which the input file resides
+* `$CPPDEP_OUTPUT_DIR`: the path to the directory in which output files are expected to be written
+* `$CPPDEP_INPUT_FILE`: the path to the file which should be tranformed by this generator command
+* `$CPPDEP_OUTPUT_PREFIX`: the path to the output file minus the file extension. This is `$CPPDEP_OUTPUT_DIR` followed by the base of the input file with its extension removed.
+
+Output files are expected to be written to `$CPPDEP_OUTPUT_PREFIX` followed by their extensions. If after running the command the expected files are not in the output directory, it is considered a failure.
+
+### Shell Generators
+Shell generators are a much more explicit version of generators. Input files and output files are explicitly defined and the generation process is controlled by running a single shell script. There are three configuration keys to be set:
+
+* `inputpaths`: `array of string` -- relative paths (relative to srcdir) to all files that will be used to create the output files
+* `outputfiles`: `array of string` -- filenames of all the file that will be created by this generator
+* `path`: `string` -- relative path to shell script
+
+#### Further Details
+The environment variable defined in the Type Generators section are all available for use within the shell script, although in practice only `$CPPDEP_OUTPUT_DIR` will be useful. All output files are expected to be written to `$CPPDEP_OUTPUT_DIR`.
+
+It is guaranteed that the shell script will be executed from the directory in which it resides, so input files can be referred to more easily with known relative paths.
 
 ## TODO
+* sort file before compiling for more consitent looking compile
 * ability to print a list of dependencies rather than compile
+* implement mode that will watch files and recompile necessary binaries when their dependencies are modified
```

Comment:
TODOs look a little tacky in a README I think, I would consider making them github issues.